packages/sdk/connect/src/connection.ts

```diff
@@ -0,0 +1,403 @@
+    this.currencyGasPrice.set(address, gasPrice)
+  }
+
+  async estimateGas(
+    tx: CeloTx,
+    gasEstimator: (tx: CeloTx) => Promise<number> = this.web3.eth.estimateGas,
+    caller: (tx: CeloTx) => Promise<string> = this.web3.eth.call
+  ): Promise<number> {
+    try {
+      const gas = await gasEstimator({ ...tx })
+      debugGasEstimation('estimatedGas: %s', gas.toString())
+      return gas
+    } catch (e) {
+      const called = await caller({ data: tx.data, to: tx.to, from: tx.from })
+      let revertReason = 'Could not decode transaction failure reason'
+      if (called.startsWith('0x08c379a')) {
+        revertReason = decodeStringParameter(this.getAbiCoder(), called.substring(10))
+      }
+      debugGasEstimation('Recover transaction failure reason', {
+        called,
+        data: tx.data,
+        to: tx.to,
+        from: tx.from,
+        error: e,
+        revertReason,
+      })
+      return Promise.reject(`Gas estimation failed: ${revertReason} or ${e}`)
+    }
+  }
+
+  getAbiCoder(): AbiCoder {
+    return (this.web3.eth.abi as unknown) as AbiCoder
+  }
+
+  async estimateGasWithInflationFactor(
+    tx: CeloTx,
+    gasEstimator?: (tx: CeloTx) => Promise<number>,
+    caller?: (tx: CeloTx) => Promise<string>
+  ): Promise<number> {
+    try {
+      const gas = Math.round(
+        (await this.estimateGas(tx, gasEstimator, caller)) * this.config.gasInflationFactor
+      )
+      debugGasEstimation('estimatedGasWithInflationFactor: %s', gas)
+      return gas
+    } catch (e) {
+      throw new Error(e)
+    }
+  }
+
+  async chainId(): Promise<number> {
+    // Reference: https://github.com/ethereum/wiki/wiki/JSON-RPC#net_version
+    const result = await this.rpcCaller.call('net_version', [])
+    return parseInt(result.result.toString(), 10)
+  }
+
+  async getTransactionCount(address: Address): Promise<number> {
+    // Reference: https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_gettransactioncount
+    const result = await this.rpcCaller.call('eth_getTransactionCount', [address, 'pending'])
+
+    const nonce = parseInt(result.result.toString(), 16)
+    return nonce
+  }
+
+  async nonce(address: Address): Promise<number> {
+    return this.getTransactionCount(address)
+  }
+
+  async coinbase(): Promise<string> {
+    // Reference: https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_coinbase
+    const result = await this.rpcCaller.call('eth_coinbase', [])
+    return result.result.toString()
+  }
+
+  async gasPrice(feeCurrency?: Address): Promise<string> {
+    // Required otherwise is not backward compatible
+    const parameter = feeCurrency ? [feeCurrency] : []
+
+    // Reference: https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_gasprice
+    const response = await this.rpcCaller.call('eth_gasPrice', parameter)
+    const gasPriceInHex = response.result.toString()
+    return gasPriceInHex
+  }
+
+  async getBlockNumber(): Promise<number> {
```

Comment:
i like how we replaced `web3` with `rpcCaller` for other methods. I would like this and the methosds below to also be replaced. Basically you need an rpcCaller (or something on top) that is able to parse inputs and parse outputs to proper types; probably more outputs that inputs). 

but it's a good idea to decouple from web3